index.js

```diff
@@ -15,9 +15,18 @@ module.exports = async function serve (drive, opts = {}) {
       return
     }
 
-    const { pathname } = new URL(req.url, 'http://localhost')
+    const { pathname, searchParams } = new URL(req.url, 'http://localhost')
+    const version = searchParams.get('checkout')
+    const driveForReq = version ? drive.checkout(version) : drive
     const filename = decodeURI(pathname)
-    const entry = await drive.entry(filename)
+
+    let entry
+    try {
+      entry = await driveForReq.entry(filename)
+    } catch (e) {
+      res.writeHead(404, 'Version not availablle').end()
```

Comment:
`availablle` -> `available`
What is the exact error for when the checkout version is not available?

Reply:
Otherwise we're replying 404 for any error

Reply:
Yeah, good point, I'll filter it out

Reply:
```
    try {
      entry = await snapshot.entry(filename)
    } catch (e) {
      if (/SNAPSHOT_NOT_AVAILABLE.*/.test(e.message)) {
        res.writeHead(404, 'Version not available').end()
        return
      }
      throw e
    }
```

Is that the ideomatic way? (so assuming those error codes are LTS)

Reply:
Probably just `err.code === 'SNAPSHOT_NOT_AVAILABLE'`

Reply:
For reference: https://github.com/holepunchto/hypercore/blob/master/lib/errors.js

Reply:
Hah, good that I checked, I already found it suspiciously complicated

Reply:
I think it would be appropriate if we return status 500 + err.message instead of throwing on other errors like when it's not `SNAPSHOT_NOT_AVAILABLE`

Reply:
Or maybe `err.code + ': ' + err.message`?

Reply:
Just message is also fine, as code could not exists, who knows haha

Reply:
`err.code || err.message` so it respond with the code if exists

Reply:
I've never used the http module directly, but won't it return a 500 status code anyway on an error? That's what most servers do (from my experience, often they return the error message in DEV mode to facilitate debugging, but not in production mode)
500 basically means: bug in code (and the server admin will filter on 500 errors so the can be fixed)

Since this is localhost, there's not really a difference between the server owner and the client. So I'm fine with adding the message as you propose.
Another option is setting up logging at the server side and returning a 500 without message to the client. Would require setting up logging already though

Reply:
Ah, but this server could also be running as an actual server right? Could be a fun way to deliver content, set up a server-drive  of a huge hyperdrive and just load content as it is requested

Reply:
But anyway, no need to worry about that yet I suppose

Reply:
(all that to say that I changed it to what you proposed, it makes sense)

Reply:
.end() + return is missing

Reply:
Very true, with the `opts.server` could easily be a https server with ssl, etc so it can serve publicly

Reply:
If it's a public server, I think showing the error code or message is not leaking any sensible information, so it should be ok

Reply:
I think `express` or libs like that would automatically handle basic errors, but unsure about `http` module

Reply:
Woops, fixed the bug

Reply:
Something like this so later we can easily handle more errors like timeout etc
```js
const msg = e.code || e.message
if (e.code === 'SNAPSHOT_NOT_AVAILABLE') res.writeHead(404, msg)
else res.writeHead(500, msg)
res.end()
return
```

Reply:
Just checked, the server crashes on an uncaught error, so that's bad.

Easy solution is to try-catch everything in the request handler, or perhaps simply refactoring at this stage to use express or some other slightly higher-level framework

Reply:
Ah wait, if you do `writeHead(code, message)` it doesn't need end()?

Reply:
I think you do, tests hang if I remove the end()

Reply:
If there is an uncaught error at this point it should crash probably haha, we can re-consider it later

Reply:
I don't think a server should ever crash if it can be avoided tbh, but indeed it doesn't belong in this PR